### ISSUE TITLE
Minified Output as Option

### ISSUE THREAD

**Issue by andrewpurkett** [joshnuss/xml_builder#13]
It is useful in some circumstances to generate XML without indentation or new lines, as such, I propose `generate` and related methods take an optional parameter that specifies if indentation and new lines should be included in output.

Currently, the only work around is to write a complex regex that removes whitespace after `xml_builder` is called, that strips the whitespace generated by `xml_builder`, but leaves whitespace where it occurs inside of data passed into `xml_builder`.

**Comment by joshnuss** [comment]
Hey @andrewpurkett,

Totally open to supporting generation without indentation.

**Comment by am-kantox** [comment]
As soon as this PR is accepted, it might be easily extended to have indentation configurable as well (the code is prepared since I have extracted `"\t"` to the `@indent` module parameter: the simple change to `defp` reading from the config will do. The same for `@blank` and `@crlf`: that way we could achieve e.g. without an indent, but still with carriage returns mode.

I doubt it’s really needed, though. I personally need a machine-readable minified file generation in the first place.

**Comment by joshnuss** [comment]
Thanks @am-kantox!

Been thinking about how to best to accomplish this.

As you pointed out, the challenge with introducing new options (like indentation) is adding an extra parameter that must be passed thru the entire call-chain.

I'm starting to think `XmlBuilder.generate`/`XmlBuilder.doc` should accept `opts` for situations like this. 

For example:

```elixir
# output with no formatting
opts = [format: XmlBuilder.Format.None]
XmlBuilder.generate(...., opts)

# output with indentation (default option)
opts = [format: XmlBuilder.Format.Indented]
XmlBuilder.generate(...., opts)
```

The formatting module would contain a `format` function (which is a copy of the current `XmlBuilder.format`)

**Comment by am-kantox** [comment]
At the very moment I am positive that configuration option as in the PR #13 should be sufficiently enough. While the `opts` parameter seems to be more handy, it’s rarely needed to change the formatting on the fly: either one has to generate human-readable, or the machine-readable stuff.
And the config option solves this task well.

The idea with `opts`, on the other hand, seems to be a good next step.

**Comment by joshnuss** [comment]
Let me think about it.

I think it might be time to switch to using `structs` internally, would clean up the code a lot and make `opts`, easier to pass around. But it might introduce breaking changes.

**Comment by am-kantox** [comment]
While you are thinking about I have refactored the code to:

- allow different formatters to be applied via [`@behaviour`](https://hexdocs.pm/elixir/Behaviour.html) implementations;
- represent an element as a `struct` (I am unsure this is needed, but since you mentioned it, I have it done);
- derive the formatting via `__using__` default formatter;
- still allow the default behaviour to be specified via config.

I strongly encourage you not to break the backward compatibility for no reason: I believe behaviours do what is needed here.

The only thing left up to you is `XmlBuilder.doc/{1,2,3}`, which is still not accepting a formatter parameter. I did not make any changes there since it’s the breaking change and whether you are still to make it, it’s definitely the maintainer’s decision on how to accomplish it better.

Enjoy!
